python/tests/test_writer.py

```diff
@@ -496,3 +496,218 @@ def test_writer_with_options(tmp_path: pathlib.Path):
+                    "val": pa.array([3, 3], pa.int64()),
+                }
+            ),
+            "expected_case2": pa.table(
+                {
+                    "p1": pa.array(["1", "1", "2", "2"], pa.string()),
+                    "p2": pa.array([False, True, False, True], pa.bool_()),
+                    "val": pa.array([2, 3, 1, 3], pa.int64()),
+                }
+            ),
+        },
+        {
+            "init": pa.table(
+                {
+                    "p1": pa.array(["1", "1", "2", "2"], pa.string()),
+                    "p2": pa.array(
+                        [
+                            date(2022, 1, 1),
+                            date(2022, 1, 2),
+                            date(2022, 1, 1),
+                            date(2022, 1, 2),
+                        ],
+                        pa.date32(),
+                    ),
+                    "val": pa.array([1, 1, 1, 1], pa.int64()),
+                }
+            ),
+            "filter1": [("p1", "=", "1")],
+            "data_case1": pa.table(
+                {
+                    "p1": pa.array(["1", "1"], pa.string()),
+                    "p2": pa.array([date(2022, 1, 1), date(2022, 1, 2)], pa.date32()),
+                    "val": pa.array([2, 2], pa.int64()),
+                }
+            ),
+            "expected_case1": pa.table(
+                {
+                    "p1": pa.array(["1", "1", "2", "2"], pa.string()),
+                    "p2": pa.array(
+                        [
+                            date(2022, 1, 1),
+                            date(2022, 1, 2),
+                            date(2022, 1, 1),
+                            date(2022, 1, 2),
+                        ],
+                        pa.date32(),
+                    ),
+                    "val": pa.array([2, 2, 1, 1], pa.int64()),
+                }
+            ),
+            "filter2": [("p2", ">", "2022-01-01")],
+            "data_case2": pa.table(
+                {
+                    "p1": pa.array(["1", "2"], pa.string()),
+                    "p2": pa.array([date(2022, 1, 2), date(2022, 1, 2)], pa.date32()),
+                    "val": pa.array([3, 3], pa.int64()),
+                }
+            ),
+            "expected_case2": pa.table(
+                {
+                    "p1": pa.array(["1", "1", "2", "2"], pa.string()),
+                    "p2": pa.array(
+                        [
+                            date(2022, 1, 1),
+                            date(2022, 1, 2),
+                            date(2022, 1, 1),
+                            date(2022, 1, 2),
+                        ],
+                        pa.date32(),
+                    ),
+                    "val": pa.array([2, 3, 1, 3], pa.int64()),
+                }
+            ),
+        },
+    ],
+)
+def test_partition_overwrite(tmp_path: pathlib.Path, cases_data: Dict[str, pa.Array]):
```

Comment:
The type for cases_data doesn't seem correct, since there is also tables and filters in there.